### DISCUSSION TITLE
Rilis Baru v1.0-RC9: Batch Messages, Scheduled Message dan lainnya

### DISCUSSION THREAD

**Discussion by rioastamal** [kirimwa-id/forum#61] [Pengumuman]
**UPDATE 31-Oktober-2021 01:25**: Bug incoming_message telah diperbaiki di v1.0-RC10.
**UPDATE 30-Oktober-2021 17:55**: Sementara rilis di-rollback ke v1.0-RC8 karena ada laporan bug incoming_message tidak masuk.

Halo semua,

Diakhir pekan bulan Oktober 2021 ini saya dengan senang hati mengumumkan dirilisnya API KirimWA.id v1.0-RC9. Fitur-fitur utama yang dihadirkan pada versi kali ini adalah **Batch Messages**, **Scheduled Messages** dan **Webhook device status**.

## Fitur Batch Messages

Dengan fitur ini developer dapat mengirimkan banyak pesan sekaligus dalam sekali request. Maksimal sebuah pesan pada sebuah batch messages adalah `50 pesan` per request.

Endpoint baru untuk fitur ini adalah `POST /v1/batch-messages`.

```sh
curl -X POST 'https://api.kirimwa.id/v1/batch-messages' \
  -H 'Authorization: Bearer API_TOKEN' \
  -H 'Content-Type: application/json' \
  -d '
{
  "messages": [
    {
      "message": "Hello from api.kirimwa.id - #1",
      "phone_number": "6281234567890",
      "message_type": "text"
    },
    {
      "message": "https://rioastamal.net/portfolio/img/rioastamal.jpg",
      "caption": "Hello from api.kirimwa.id - #2",
      "phone_number": "6281234567890-1628639662",
      "is_group_message": true,
      "message_type": "image",
      "send_at": "2021-10-17T06:50:05+07:00"
    }
  ],
  "device_id": "iphone-x-pro"
}
'
```

Dokumentasi lengkap ada di https://developer.kirimwa.id/api/v1/#batch-messages.

## Fitur Scheduled Message

Dengan fitur ini developer dapat menenukan tanggal dan waktu pesan akan dikirimkan.  API KirimWA.id akan otomatis mengatur penjadwalan pengiriman tersebut sehingga developer tidak perlu melakukannya sendiri misal lewat cron jobs.

Atribut yang digunakan untuk menentukan jadwal pesan dikirim adalah `send_at`. Atribut ini bersifat opsional, jika tidak diisi maka akan dianggap bernilai `now` dan pesan langsung dikirim.

```sh
curl -X POST 'https://api.kirimwa.id/v1/messages' \
  -H 'Authorization: Bearer API_TOKEN' \
  -H 'Content-Type: application/json' \
  -d '
{
  "phone_number": "6281234567890",
  "message": "Halo ini adalah pesan dari api.kirimwa.id",
  "device_id": "iphone-x-pro",
  "message_type": "text",
  "send_at": "2021-10-15T14:30:00+07:00"
}
'
```

Dokumentasi lengkap ada di https://developer.kirimwa.id/api/v1/#kirim-pesan-terjadwal.

## Fitur Webhook Device Status

Sekarang ketika status koneksi dari device berubah misal dari connected ke disconnected atau sebaliknya API KirimWA.id akan mengirimkan status tersebut ke webhook yang telah didaftarkan. Nilai dari `webhook_type` untuk notifikasi ini adalah `device_status_changed`.

Berikut ini adalah salah satu contoh payload webhook yang dikirimkan.

```json
{
  "webhook_type": "device_status_changed",
  "device_id": "asus-zenfone-1",
  "status": "connected",
  "changed_at": "2021-10-30T02:03:23.617Z",
  "server_time": "2021-10-30T02:03:23.624Z"
}
```

Dokumentasi lengkap ada di https://developer.kirimwa.id/api/v1/#webhook-device-status.

Saya harap dengan fitur ini akan mempermudah teman-teman developer dalam mengembangkan aplikasi yang mengandalkan API KirimWA.id.

Belum memiliki Token API KirimWA.id? Daftar gratis di https://developer.kirimwa.id/

Salam,
Rio Astamal

**Comment by fahram** [comment]
Bang Rio. 
setelah release ini saya tidak dapat lagi   'webhook_type' => 'incoming_message'

**Reply by rioastamal** [reply]
Terima kasih laporannya mas @fahram, bug confirmed. Rilis telah di rollback ke v1.0-RC8 dan akan saya investigasi nanti.

**Reply by fahram** [reply]
Siap bang @rioastamal , terimakasih banyak

**Reply by rioastamal** [reply]
Silahkan dicoba kembali mas @fahram

**Reply by fahram** [reply]
sudah dicoba bang, sudah OK 
Terimakasih.
